flocker/control/_protocol.py

```diff
@@ -123,46 +124,27 @@ def fromBox(self, name, strings, objects, proto):
         self.another_argument.fromBox(name, strings, objects, proto)
 
 
-class CachingEncoder(object):
-    """
-    Cache results of ``wire_encode`` and re-use them, relying on the fact
-    we're encoding immutable objects.
+# The configuration and state can get pretty big, so don't want too many:
+_wire_encode_cache = LRUCache(50)
 
-    Not thread-safe, so should only be used by a single thread (the
-    Twisted reactor thread, presumably).
 
-    :attr _cache: Either ``None`` indicating no caching or a dicitonary
-        with objects mapped to cached wire encoded values.
+def caching_wire_encode(obj):
     """
-    def __init__(self):
-        self._cache = None
-
-    def encode(self, obj):
-        """
-        Encode an object to bytes using ``wire_encode``, or return cached
-        result if available and running in context of ``cache()`` context
-        manager.
-
-        :param obj: Object to encode.
-        :return: Resulting ``bytes``.
-        """
-        if self._cache is None:
-            return wire_encode(obj)
+    Encode an object to bytes using ``wire_encode`` and cache the result,
+    or return cached result if available.
 
-        if obj not in self._cache:
-            self._cache[obj] = wire_encode(obj)
-        return self._cache[obj]
+    This relies on cached objects being immutable, or at least not being
+    modified. Given our usage patterns that is currently the case and
+    should continue to be, but worth keeping in mind.
 
-    @contextmanager
-    def cache(self):
-        """
-        While in context of this context manager results will be cached.
-        """
-        self._cache = {}
-        yield
-        self._cache = None
-
-_caching_encoder = CachingEncoder()
+    :param obj: Object to encode.
+    :return: Resulting ``bytes``.
+    """
+    result = _wire_encode_cache.get(obj)
+    if result is None:
+        result = wire_encode(obj)
+        _wire_encode_cache.put(obj, result)
```

Comment:
As you've noted, this relies on the cached objects' immutability. That makes this look a little bit flaky, but actually we already rely on the type of objects we serialize being immutable in our existing context cache, which I think are mostly ``Deployment`` and maybe one or two other similar models anyway.